### PR TITLE
sql: begin to use structured errors in purification

### PR DESCRIPTION
**Pull request by guswynn**
https://github.com/MaterializeInc/materialize/issues/15605 is evidence that users are beginning to struggle to comprehend at least some of the error we give them in sql. This is partially because of the generic `anyhow::Error` we use in purification. This pr changes the purification crate to use structured errors, and additionally improves the postgres errors (by clearly indicating that the error came from talking to postgres), which appears to be the most heavily-misunderstood error we emit.

In the future, we may want to better capture the `chain` of errors that cause an error, but for now, the display impls seem sufficient. Also note that this pr does not change any error message other than postgres ones, but setups a structure that makes improving these in the future better. Ideally, this structure would also be reflected in our error logs within storaged (and other services), but one step at a time.

### Motivation

  * This PR fixes a recognized bug.
Partially resolves: https://github.com/MaterializeInc/materialize/issues/15605 (this codepath is also not tested, as i couldn't figure out how to trigger a timeout in ci (cc @philip-stoev )

### Tips for reviewer

The bottom commit just adds structure to errors in purification. The second commit adds structure to the underlying postgres errors, and attempts to better format

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

File: src/sql/src/pure.rs [hunk 1]
```diff
@@ -55,10 +55,58 @@ use crate::plan::error::PlanError;
...
+}
+
+macro_rules! bail_generic {
+    ($fmt:expr, $($arg:tt)*) => {
+        return Err(PurificationError::Generic(anyhow::anyhow!($fmt, $($arg)*)))
+    };
+    ($err:expr $(,)?) => {
+        return Err(PurificationError::Generic(anyhow::anyhow!($err)))
+    };
+}
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/sql/src/pure.rs:104]: This has to return a `PurificationError` tho?

File: src/sql/src/pure.rs [hunk 2]
```diff
@@ -55,10 +55,58 @@ use crate::plan::error::PlanError;
 use crate::plan::statement::ddl::load_generator_ast_to_generator;
 use crate::plan::StatementContext;
 
+/// An error that happened during purification.
+/// This should be comprehensible (and actionable)
+/// by a user.
+#[derive(Debug, thiserror::Error)]
+pub enum PurificationError {
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/sql/src/pure.rs:62]: What about using `PurificationError` as a variant of `PlanError`?

[the remark this answers is not in the document]
Comment (benesch) [reply src/sql/src/pure.rs:62]: I don't think purification errors are a strict subset of plan errors! There are several errors that are shared between the two.

**Comment by philip-stoev** [comment]
@guswynn attempting to connect to a non-routable IP, e.g. 203.0.113.0 should eventually result in a timeout. Not sure however how long we will wait so it may not be suitable for a CI test.

File: src/sql/src/pure.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (guswynn) [reply src/sql/src/pure.rs:62]: So I should add `tokio_postgres::Error` as a variant of `PlanError`?

**Comment by guswynn** [comment]
@philip-stoev I can try it out, i think its 2minutes typically

File: src/postgres-util/src/lib.rs [hunk 3]
```diff
@@ -29,8 +28,39 @@ use crate::desc::{PostgresColumnDesc, PostgresTableDesc};
 
 pub mod desc;
 
+/// An error that happened during purification.
+/// This should be comprehensible (and actionable)
+/// by a user.
```
Comment (benesch) [inline src/postgres-util/src/lib.rs:33]: I think this is copypasta from an earlier version of this PR?

**Review by benesch — COMMENTED** [review]
Thanks, yeah, I like where this is headed! Here's my vision for what the `PlanError` error variants should look like:

```diff
diff --git a/src/sql/src/error_formatting.rs b/src/sql/src/error_formatting.rs
deleted file mode 100644
index d05a3023c..000000000
--- a/src/sql/src/error_formatting.rs
+++ /dev/null
@@ -1,53 +0,0 @@
-// Copyright Materialize, Inc. and contributors. All rights reserved.
-//
-// Use of this software is governed by the Business Source License
-// included in the LICENSE file.
-//
-// As of the Change Date specified in that file, in accordance with
-// the Business Source License, use of this software will be governed
-// by the Apache License, Version 2.0.
-
-//! This module contains functions that format errors into meaningful
-//! messages for users of materialize's sql. They are intended to be
-//! used in `Display` impl's in structured error-types in this crate.
-
-use std::error::Error;
-use std::fmt;
-
-pub trait UserFacingError {
-    /// The main error displayed to the user.
-    fn user_fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error>;
-    /// The `DETAIL` displayed to the user.
-    fn detail(&self) -> String;
-    /// The optional `HINT` displayed to the user.
-    fn hint(&self) -> Option<String>;
-}
-
-impl UserFacingError for tokio_postgres::Error {
-    fn user_fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
-        write!(f, "error communicating with postgres instance")
-    }
-
-    fn detail(&self) -> String {
-        format!("{}", self)
-    }
-
-    fn hint(&self) -> Option<String> {
-        // unfortunately we need a triple-nested if here...
-        if let Some(cause) = self.source() {
-            if let Some(io_err) = cause.downcast_ref::<std::io::Error>() {
-                // TODO(guswynn): add timeout value, and link to more docs
-                if io_err.kind() == std::io::ErrorKind::TimedOut {
-                    return Some(
-                        "Timeouts can sometimes indicate \
-                        a firewall. Does materialize have access to \
-                        the postgres instance?"
-                            .to_string(),
-                    );
-                }
-            }
-        }
-
-        None
-    }
-}
diff --git a/src/sql/src/lib.rs b/src/sql/src/lib.rs
index 62c444e2c..1ea9ca5f8 100644
--- a/src/sql/src/lib.rs
+++ b/src/sql/src/lib.rs
@@ -98,7 +98,6 @@ pub const DEFAULT_SCHEMA: &str = "public";
 
 pub mod ast;
 pub mod catalog;
-mod error_formatting;
 pub mod func;
 pub mod kafka_util;
 pub mod names;
diff --git a/src/sql/src/plan/error.rs b/src/sql/src/plan/error.rs
index 34faaffcd..436f00743 100644
--- a/src/sql/src/plan/error.rs
+++ b/src/sql/src/plan/error.rs
@@ -9,6 +9,7 @@
 
 use std::error::Error;
 use std::fmt;
+use std::io;
 use std::num::ParseIntError;
 use std::num::TryFromIntError;
 use std::sync::Arc;
@@ -25,7 +26,6 @@ use mz_repr::ColumnName;
 use mz_sql_parser::parser::ParserError;
 
 use crate::catalog::CatalogError;
-use crate::error_formatting::UserFacingError;
 use crate::names::PartialObjectName;
 use crate::names::ResolvedObjectName;
 use crate::plan::plan_utils::JoinSide;
@@ -93,9 +93,16 @@ pub enum PlanError {
         column: String,
         e: TypeFromOidError,
     },
-    CsrGetSchema(Arc<crate::pure::GetSchemaError>),
-    Protobuf(protobuf_native::OperationFailedError),
-    Postgres(Arc<tokio_postgres::Error>),
+    FetchingCsrSchemaFailed {
+        schema_lookup: String,
+        cause: Arc<dyn Error + Send + Sync>,
+    },
+    FetchingPostgresPublicationInfoFailed {
+        cause: Arc<mz_postgres_util::PostgresError>,
+    },
+    InvalidProtobufSchema {
+        cause: protobuf_native::OperationFailedError,
+    },
     // TODO(benesch): eventually all errors should be structured.
     Unstructured(String),
 }
@@ -115,7 +122,11 @@ impl PlanError {
                 column: _,
                 e,
             } => Some(format!("{e}")),
-            Self::Postgres(e) => Some(e.detail()),
+            Self::FetchingCsrSchemaFailed { cause, .. } => Some(cause.to_string()),
+            Self::FetchingPostgresPublicationInfoFailed { cause } => {
+                Some(cause.to_string())
+            }
+            Self::InvalidProtobufSchema { cause } => Some(cause.to_string()),
             _ => None,
         }
     }
@@ -140,7 +151,6 @@ impl PlanError {
             Self::UnacceptableTimelineName(_) => {
                 Some("The prefix \"mz_\" is reserved for system timelines.".into())
             }
-            Self::Postgres(e) => e.hint(),
             Self::UnrecognizedTypeInPostgresSource {
                 table: _,
                 column: _,
@@ -150,6 +160,20 @@ impl PlanError {
                 Try excluding the table from the publication."
                     .into(),
             ),
+            Self::FetchingPostgresPublicationInfoFailed { cause } => {
+                if let Some(cause) = cause.source() {
+                    if let Some(cause) = cause.downcast_ref::<io::Error>() {
+                        if cause.kind() == io::ErrorKind::TimedOut {
+                            return Some(
+                                "Do you have a firewall or security group that is \
+                                preventing Materialize from conecting to your PostgreSQL server?"
+                                    .into(),
+                            );
+                        }
+                    }
+                }
+                None
+            }
             _ => None,
         }
     }
@@ -253,10 +277,14 @@ impl fmt::Display for PlanError {
                 "column {} uses unrecognized type",
                 format!("{}.{}", table, column).quoted()
             ),
-            Self::CsrGetSchema(e) => e.fmt(f),
-            Self::Protobuf(e) => e.fmt(f),
-            Self::Postgres(e) => {
-                e.user_fmt(f)
+            Self::FetchingCsrSchemaFailed { schema_lookup, .. } => {
+                write!(f, "failed to fetch schema {schema_lookup} from schema registry")
+            }
+            Self::FetchingPostgresPublicationInfoFailed { .. } => {
+                write!(f, "failed to fetch publication information from PostgreSQL database")
+            }
+            Self::InvalidProtobufSchema { .. } => {
+                write!(f, "invalid protobuf schema")
             }
             Self::DropSubsource{subsource, source: _} => write!(f, "SOURCE {subsource} is a subsource and cannot be dropped independently of its primary source"),
         }
@@ -331,36 +359,6 @@ impl From<ParserError> for PlanError {
     }
 }
 
-impl From<mz_ccsr::GetBySubjectError> for PlanError {
-    fn from(e: mz_ccsr::GetBySubjectError) -> PlanError {
-        PlanError::CsrGetSchema(Arc::new(crate::pure::GetSchemaError::Subject(e)))
-    }
-}
-
-impl From<crate::pure::GetSchemaError> for PlanError {
-    fn from(e: crate::pure::GetSchemaError) -> PlanError {
-        PlanError::CsrGetSchema(Arc::new(e))
-    }
-}
-
-impl From<protobuf_native::OperationFailedError> for PlanError {
-    fn from(e: protobuf_native::OperationFailedError) -> PlanError {
-        PlanError::Protobuf(e)
-    }
-}
-
-// TODO(guswynn): determine if `Clone` is required on `PlanError`
-impl From<mz_postgres_util::PostgresError> for PlanError {
-    fn from(e: mz_postgres_util::PostgresError) -> Self {
-        use mz_postgres_util::PostgresError::*;
-        match e {
-            Postgres(e) => PlanError::Postgres(Arc::new(e)),
-            // TODO(guswynn): better categorize these
-            other => sql_err!("{}", other),
-        }
-    }
-}
-
 impl From<QGMError> for PlanError {
     fn from(e: QGMError) -> PlanError {
         PlanError::Qgm(e)
diff --git a/src/sql/src/pure.rs b/src/sql/src/pure.rs
index c51796b3f..68a532ff5 100644
--- a/src/sql/src/pure.rs
+++ b/src/sql/src/pure.rs
@@ -12,7 +12,6 @@
 //! See the [crate-level documentation](crate) for details.
 
 use std::collections::HashMap;
-use std::error::Error as StdError;
 use std::iter;
 use std::path::Path;
 use std::sync::Arc;
@@ -29,6 +28,7 @@ use mz_ccsr::Schema as CcsrSchema;
 use mz_ccsr::{Client, GetByIdError, GetBySubjectError};
 use mz_cloud_resources::AwsExternalIdPrefix;
 use mz_ore::cast::CastFrom;
+use mz_ore::str::StrExt;
 use mz_proto::RustType;
 use mz_repr::{strconv, GlobalId};
 use mz_secrets::SecretsReader;
@@ -294,7 +294,11 @@ pub async fn purify_create_source(
             let config = connection
                 .config(&*connection_context.secrets_reader)
                 .await?;
-            let tables = mz_postgres_util::publication_info(&config, &publication).await?;
+            let tables = mz_postgres_util::publication_info(&config, &publication)
+                .await
+                .map_err(|cause| PlanError::FetchingPostgresPublicationInfoFailed {
+                    cause: Arc::new(cause),
+                })?;
 
             let mut targeted_subsources = vec![];
 
@@ -765,60 +769,30 @@ pub struct Schema {
     pub value_schema: String,
 }
 
-#[derive(Debug)]
-pub enum GetSchemaError {
-    Subject(GetBySubjectError),
-    Id(GetByIdError),
-}
-
-impl From<GetBySubjectError> for GetSchemaError {
-    fn from(inner: GetBySubjectError) -> Self {
-        Self::Subject(inner)
-    }
-}
-
-impl From<GetByIdError> for GetSchemaError {
-    fn from(inner: GetByIdError) -> Self {
-        Self::Id(inner)
-    }
-}
-
-impl std::fmt::Display for GetSchemaError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        match self {
-            GetSchemaError::Subject(e) => write!(f, "failed to look up schema by subject: {e}"),
-            GetSchemaError::Id(e) => write!(f, "failed to look up schema by id: {e}"),
-        }
-    }
-}
-
-impl StdError for GetSchemaError {
-    fn source(&self) -> Option<&(dyn StdError + 'static)> {
-        match self {
-            GetSchemaError::Subject(e) => Some(e),
-            GetSchemaError::Id(e) => Some(e),
-        }
-    }
-}
-
 async fn get_schema_with_strategy(
     client: &Client,
     strategy: ReaderSchemaSelectionStrategy,
     subject: &str,
-) -> Result<Option<String>, GetSchemaError> {
+) -> Result<Option<String>, PlanError> {
     match strategy {
         ReaderSchemaSelectionStrategy::Latest => {
             match client.get_schema_by_subject(subject).await {
                 Ok(CcsrSchema { raw, .. }) => Ok(Some(raw)),
                 Err(GetBySubjectError::SubjectNotFound) => Ok(None),
-                Err(e) => Err(e.into()),
+                Err(e) => Err(PlanError::FetchingCsrSchemaFailed {
+                    schema_lookup: format!("subject {}", subject.quoted()),
+                    cause: Arc::new(e),
+                }),
             }
         }
         ReaderSchemaSelectionStrategy::Inline(raw) => Ok(Some(raw)),
         ReaderSchemaSelectionStrategy::ById(id) => match client.get_schema_by_id(id).await {
             Ok(CcsrSchema { raw, .. }) => Ok(Some(raw)),
             Err(GetByIdError::SchemaNotFound) => Ok(None),
-            Err(e) => Err(e.into()),
+            Err(e) => Err(PlanError::FetchingCsrSchemaFailed {
+                schema_lookup: format!("ID {}", id),
+                cause: Arc::new(e),
+            }),
         },
     }
 }
@@ -830,14 +804,8 @@ async fn get_remote_csr_schema(
     topic: String,
 ) -> Result<Schema, PlanError> {
     let value_schema_name = format!("{}-value", topic);
-    let value_schema = get_schema_with_strategy(ccsr_client, value_strategy, &value_schema_name)
-        .await
-        .with_context(|| {
-            format!(
-                "fetching latest schema for subject '{}' from registry",
-                value_schema_name
-            )
-        })?;
+    let value_schema =
+        get_schema_with_strategy(ccsr_client, value_strategy, &value_schema_name).await?;
     let value_schema = value_schema.ok_or_else(|| anyhow!("No value schema found"))?;
     let subject = format!("{}-key", topic);
     let key_schema = get_schema_with_strategy(ccsr_client, key_strategy, &subject).await?;
@@ -852,8 +820,13 @@ async fn compile_proto(
     subject_name: &String,
     ccsr_client: &Client,
 ) -> Result<CsrSeedProtobufSchema, PlanError> {
-    let (primary_subject, dependency_subjects) =
-        ccsr_client.get_subject_and_references(subject_name).await?;
+    let (primary_subject, dependency_subjects) = ccsr_client
+        .get_subject_and_references(subject_name)
+        .await
+        .map_err(|e| PlanError::FetchingCsrSchemaFailed {
+            schema_lookup: format!("subject {}", subject_name.quoted()),
+            cause: Arc::new(e),
+        })?;
 
     // Compile .proto files into a file descriptor set.
     let mut source_tree = VirtualSourceTree::new();
@@ -866,7 +839,8 @@ async fn compile_proto(
     let mut db = SourceTreeDescriptorDatabase::new(source_tree.as_mut());
     let fds = db
         .as_mut()
-        .build_file_descriptor_set(&[Path::new(&primary_subject.name)])?;
+        .build_file_descriptor_set(&[Path::new(&primary_subject.name)])
+        .map_err(|cause| PlanError::InvalidProtobufSchema { cause })?;
 
     // Ensure there is exactly one message in the file.
     let primary_fd = fds.file(0);
@@ -877,8 +851,11 @@ async fn compile_proto(
     };
 
     // Encode the file descriptor set into a SQL byte string.
+    let bytes = &fds
+        .serialize()
+        .map_err(|cause| PlanError::InvalidProtobufSchema { cause })?;
     let mut schema = String::new();
-    strconv::format_bytes(&mut schema, &fds.serialize()?);
+    strconv::format_bytes(&mut schema, bytes);
 
     Ok(CsrSeedProtobufSchema {
         schema,
```

The salient point is that I'd like to get away from the `From` impls. They encourage folks to just use `?` willy nilly without adding the appropriate context. Adding the explicit error variants for each different place the error is thrown is a bit more painful but makes the error messages _much_ better. IMO the only place we should be using `From` impls for user-facing error messages is with error types that are already expecting to be user facing—i.e., error variants that come with `hint` and `detail` inherent methods.

File: src/postgres-util/src/lib.rs (hunk 3 above)
Comment (guswynn) [reply src/postgres-util/src/lib.rs:33]: ugh, sorry about that!

**Comment by guswynn** [comment]
@benesch removing the `From` impls seems reasonable to me!

I think at some point we will want to move the `hint` and `detail` impls into their own module, as we produce more and more specific errors for as many things as possible

this should be ready to review (and hopefully passing ci); In the future asserting stuff about `detail`'s and `hint`'s in td would be cool, but im unsure of the syntax i would want for it

**Comment by guswynn** [comment]
@benesch @philip-stoev stacked pr with td hint/detail assertions: https://github.com/MaterializeInc/materialize/pull/16030

Can we merge this pr (it reflects all your above comments and suggestions) first, and then merge the stacked pr with additional tests?

**Review by benesch — APPROVED** [review]
This looks good. Thanks for iterating, @guswynn. I'm going to hit merge, since it looks like there are no conflicts now, and I'd hate for that to change before you log on tomorrow!

**Pull request merged**